### PR TITLE
sink: reduce lock competition during flushBackendSink

### PR DESCRIPTION
**Pull request by hi-rustin**
### What problem does this PR solve? 
None

### What is changed and how it works?
reduce lock competition during flushBackendSink.

### Check List 

Tests 

 -  Benchmark (TBD)

Code changes

None

Side effects

None

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
None
```

/cc

### CODE REVIEW

**Comment by hi-rustin** [comment]
/hold

- [ ] Benchmark test

**Comment by hi-rustin** [comment]
Benchmark result: 

Before:

```sh
➜  sink git:(rustin-patch-fix-sink) ✗ go test  -bench='^\QBenchmarkManagerRandomFlushing\E$' -test.run=thisexpressionwontmatchanytest -count=10 .
goos: darwin
goarch: amd64
pkg: github.com/pingcap/ticdc/cdc/sink
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkManagerRandomFlushing-16              1        4466272548 ns/op
BenchmarkManagerRandomFlushing-16              1        4692509999 ns/op
BenchmarkManagerRandomFlushing-16              1        4566925133 ns/op
BenchmarkManagerRandomFlushing-16              1        4553761846 ns/op
BenchmarkManagerRandomFlushing-16              1        4628824677 ns/op
BenchmarkManagerRandomFlushing-16              1        4725037126 ns/op
BenchmarkManagerRandomFlushing-16              1        4786872104 ns/op
BenchmarkManagerRandomFlushing-16              1        4513431907 ns/op
BenchmarkManagerRandomFlushing-16              1        4583744744 ns/op
BenchmarkManagerRandomFlushing-16              1        4616345649 ns/op
PASS
ok      github.com/pingcap/ticdc/cdc/sink       47.078s
```

After:
```sh
➜  sink git:(rustin-patch-fix-sink) ✗ go test  -bench='^\QBenchmarkManagerRandomFlushing\E$' -test.run=thisexpressionwontmatchanytest -count=10 .
goos: darwin
goarch: amd64
pkg: github.com/pingcap/ticdc/cdc/sink
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkManagerRandomFlushing-16             20          55283572 ns/op
BenchmarkManagerRandomFlushing-16             24          53067853 ns/op
BenchmarkManagerRandomFlushing-16             21          53231288 ns/op
BenchmarkManagerRandomFlushing-16             20          51749307 ns/op
BenchmarkManagerRandomFlushing-16             24          52132987 ns/op
BenchmarkManagerRandomFlushing-16             25          52443937 ns/op
BenchmarkManagerRandomFlushing-16             22          52733766 ns/op
BenchmarkManagerRandomFlushing-16             19          53113452 ns/op
BenchmarkManagerRandomFlushing-16             21          53181457 ns/op
BenchmarkManagerRandomFlushing-16             24          53821486 ns/op
PASS
ok      github.com/pingcap/ticdc/cdc/sink       13.207s
```

**Comment by hi-rustin** [comment]
/cc @overvenus @amyangfei 

Could you please take a look? Thanks!

File: cdc/sink/manager.go [hunk 1]
```diff
@@ -94,8 +95,14 @@ func (m *Manager) getMinEmittedTs() model.Ts {
 }
 
 func (m *Manager) flushBackendSink(ctx context.Context) (model.Ts, error) {
+	if !atomic.CompareAndSwapInt64(&m.flushing, 0, 1) {
+		return atomic.LoadUint64(&m.checkpointTs), nil
+	}
```
Comment (overvenus) [inline cdc/sink/manager.go:100]: Could you add some comments?

Comment (hi-rustin) [reply cdc/sink/manager.go:103]: Done.

**Review by overvenus — APPROVED** [review]
LGTM

**Comment by hi-rustin** [comment]
@amyangfei 
Could you please take a look? Thanks!

File: cdc/sink/manager.go [hunk 2]
```diff
@@ -94,8 +95,17 @@ func (m *Manager) getMinEmittedTs() model.Ts {
 }
 
 func (m *Manager) flushBackendSink(ctx context.Context) (model.Ts, error) {
+	// NOTICE: Because all table sinks will try to flush backend sink,
+	// which will cause a lot of lock contention and blocking in high concurrency cases.
+	// So here we use flushing as a lightweight lock to improve the lock competition problem.
+	if !atomic.CompareAndSwapInt64(&m.flushing, 0, 1) {
+		return atomic.LoadUint64(&m.checkpointTs), nil
```
Comment (amyangfei) [inline cdc/sink/manager.go:102]:
```suggestion
		return m.getCheckpointTs(), nil
```

**Review by amyangfei — APPROVED** [review]

**Pull request merged**